### ISSUE TITLE
i18n on validation and database errors

### ISSUE THREAD

**Issue by GTTo** [sequelize/sequelize#4361]
Hello!
When  I create a model and I define a validation over an attribute, I can define my own messages:
````
sequelize.define('Foo', {
  foo: {
    type: Sequelize.STRING,
    validate: {
      isInt: {
         msg: "Must be an integer number of pennies"
       }
    }
  }
}
````
In this case, you can translate the message as you want.

When an sql validation error is thrown, the messages is only in English
lib/data-types.js
````
INTEGER.prototype.validate = function(value) {
  if (!Validator.isInt(value)) {
    throw new sequelizeErrors.ValidationError(util.format('%j is not a valid integer', value));
  }
  return true;
};
````
How can I translate this messages? Why It doesn't have an i18n file to translate it?

Thank you!

**Comment by GTTo** [comment]
there is anyone out there?

**Comment by janmeier** [comment]
@GTTo , yes, but we all have jobs, school, family, etc. on the side - so response is not always instantaneous even though we try to be fast. :)

I'm not really sure this is a complexity we wan't to handle in sequelize. When you're at the point were your app needs i18n I think it should be handled somewhere else than directly in the ORM. 

I do realize that the particular error you're refering to is hard to translate - it's a built in validation, and has nothing to do with the `isInt` validator - see https://github.com/sequelize/sequelize/pull/3472. So the only way to find out that the validationerror comes from this validator is to do string matching, which is ugly.

We should probably add a `type` or similar to the error / use errorItem so users can check which validation actually failed

cc @mickhansen and possibly @fixe

**Comment by fixe** [comment]
Since string matching doesn't work in this case because the value is dynamic I'm +1 for providing a `type` or `code` associated with the validation error.

**Comment by billxinli** [comment]
I am facing this issue as well, application is I18N, however, I don't really want to handle translation in the model/ORM layer. I also don't want to do string matching of if err === "this message" then translate with this key

Since Sequelize can be used in conjunction with other libraries, I am wondering would it make sense to have a possible to have an abstract callback layer to handle the translations? Users now can have a method to provide a set of override to handle translations, if one isn't provided, sequelize will fallback to the default hardcoded strings?

**Comment by misterdai** [comment]
Run into the same issue as well.  I was hoping to set the error message for the validation rules as an array of arguments for the i18n module to then translate, looping over validation errors with the afterValidate hook.  But anything I seem to set against the `msg` attribute, is converted to a string.  While I can just split the string, it's not ideal and wouldn't work if one of the parameters contained a comma themselves.

Allowing the `msg` attribute to pass through, or an additional `data` attribute, would be very useful.  I'm not sure if it's sequelize or validator.js that's making the conversion between types though.

**Comment by misterdai** [comment]
My solution is currently:

1. All validation messages are a delimited string of parameters intended for the `i18n` module.  Due to array's being converted to a string after validation.
2. Added a `i18n` virtual field to models that will hold the locale required for the messages.
3. Added `validationFailed` hook.
  a. Creates an instance of `i18n` (avoiding global locale settings).
  b. Sets instance to the virtual field implied locale.
  c. Loops over validation error messages, splitting them into an array and `apply`'ing them to the `i18n` instance translation method.

While the above works, it's not pleasant and means that I'm currently stuck using a single translation method `__` as supporting others for different types of translation support would be even more messy.

**Comment by misterdai** [comment]
Still an issue, it would be fantastic to have a way of supporting translatable validation messages.
